src/lib/mdns/minimal/ActiveResolveAttempts.cpp

```diff
@@ -0,0 +1,177 @@
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ActiveResolveAttempts.h"
+
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <algorithm>
+
+using namespace chip;
+
+namespace mdns {
+namespace Minimal {
+
+void ActiveResolveAttempts::Reset()
+
+{
+    for (auto & item : mRetryQueue)
+    {
+        item.peerId.SetNodeId(kUndefinedNodeId);
+    }
+}
+
+void ActiveResolveAttempts::Complete(const PeerId & peerId)
+{
+    for (auto & item : mRetryQueue)
+    {
+        if (item.peerId == peerId)
+        {
+            item.peerId.SetNodeId(kUndefinedNodeId);
+            return;
+        }
+    }
+
+    // This may happen during boot time adverisements: nodes come online
+    // and advertise their IP without any explicit queries for them
+    ChipLogProgress(Discovery, "Discovered node without a pending query");
+}
+
+void ActiveResolveAttempts::MarkPending(const PeerId & peerId)
+{
+    // Strategy when picking the peer id to use:
+    //   1 if a matching peer id is already found, use that one
+    //   2 if an 'unused' entry is found, use that
+    //   3 otherwise expire the one with the largest nextRetryDelaySec
+    //     or if equal nextRetryDelaySec, pick the one with the oldest
+    //     queryDueTimeMs
+
+    RetryEntry * entryToUse = &mRetryQueue[0];
+
+    for (size_t i = 1; i < kRetryQueueSize; i++)
+    {
+        if (entryToUse->peerId == peerId)
+        {
+            break; // best match possible
+        }
+
+        RetryEntry * entry = mRetryQueue + i;
+
+        // Rule 1: peer id match always matches
+        if (entry->peerId == peerId)
+        {
+            entryToUse = entry;
+            continue;
+        }
+
+        // Rule 2: select unused entries
+        if ((entryToUse->peerId.GetNodeId() != kUndefinedNodeId) && (entry->peerId.GetNodeId() == kUndefinedNodeId))
+        {
+            entryToUse = entry;
+            continue;
+        }
+        else if (entryToUse->peerId.GetNodeId() == kUndefinedNodeId)
+        {
+            continue;
+        }
+
+        // Rule 2: both choices are used (have a defined node id),
+        // pick the one with most queryDueTimeMs
+        if (entry->nextRetryDelaySec > entryToUse->nextRetryDelaySec)
+        {
+            entryToUse = entry;
+        }
+        else if ((entry->nextRetryDelaySec == entryToUse->nextRetryDelaySec) &&
+                 (entry->queryDueTimeMs < entryToUse->queryDueTimeMs))
+        {
+            entryToUse = entry;
+        }
+    }
+
+    if ((entryToUse->peerId.GetNodeId() != kUndefinedNodeId) && (entryToUse->peerId != peerId))
+    {
+        ChipLogError(Discovery, "Re-using pending resolve entry before reply was received.");
```

Comment:
Don't we need an error callback here for the evicted resolution operation?

Reply:
Added a TODO - I did not attempt to add callbacks for error cases in this PR, only retries. Callbacks could/should be added for such things but I believe separately.